### PR TITLE
Bluetooth: host: Invalid role used in finding connections

### PR DESCRIPTION
**Pull request by cvinayak**
Directed advertising timeout was dropped in the host due to
not being able to find a pending connection. Host used the
role parameter from the error event parameters which has
been earlier zero-ed out.

Regression introduced in commit a0349689ff5e ("Bluetooth:
host: Fix conn object assigned to wrong connection")

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/host/hci_core.c [hunk 1]
```diff
@@ -1311,42 +1311,63 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
...
+			conn = find_pending_connect(BT_HCI_ROLE_SLAVE, NULL);
+			if (!conn) {
+				BT_ERR("No pending slave connection");
+
+				return;
+			}
+
+			conn->err = evt->status;
+
 			atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c]: I think we should take the oppurtunity to handle this as well:
```
		/* Clear advertising even if we are not able to add connection
		 * object to keep host in sync with controller state
		 */
		atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
```
Move above checking if no conn.

File: subsys/bluetooth/host/hci_core.c [hunk 2]
```diff
@@ -1311,42 +1311,63 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
...
-
 		if (IS_ENABLED(CONFIG_BT_PERIPHERAL) &&
-		    conn->err == BT_HCI_ERR_ADV_TIMEOUT) {
+		    evt->status == BT_HCI_ERR_ADV_TIMEOUT) {
 			/*
 			 * Handle advertising timeout after high duty directed
 			 * advertising.
 			 */
 
+			/*
```
Comment (joerchan) [inline subsys/bluetooth/host/hci_core.c]: Put the two comments together perhaps?

**Review by joerchan — CHANGES_REQUESTED** [review]
So that's why :) Good call to add the BT_ERR
Would you mind making sure the advertising state is correctly updated in this case?

@jhedberg Would it make sense to `BT_ASSERT(conn);` There should always be a connection object.

File: subsys/bluetooth/host/hci_core.c (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/host/hci_core.c]: placed atomic_clear_bit in between.

**Review by joerchan — APPROVED** [review]

File: subsys/bluetooth/host/hci_core.c [hunk 3]
```diff
@@ -1312,41 +1312,58 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
...
+
+			conn->err = evt->status;
+
 			le_conn_cancel_complete(conn);
 			goto done;
 		}
 
 		BT_WARN("Unexpected status 0x%02x", evt->status);
 
-		bt_conn_unref(conn);
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: Why do you remove this? Looks like it's introducing a leak of the reference returned by `find_pending_connect()`

File: subsys/bluetooth/host/hci_core.c [hunk 4]
```diff
@@ -1312,41 +1312,58 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
...
+			/*
+			 * Handle create connection cancel.
+			 *
+			 * There is no need to check ID address as only one
+			 * connection in master role can be in pending state.
+			 */
+			conn = find_pending_connect(BT_HCI_ROLE_MASTER, NULL);
+			if (!conn) {
+				BT_ERR("No pending master connection");
+
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: minor: remove this empty line

File: subsys/bluetooth/host/hci_core.c [hunk 5]
```diff
@@ -1312,41 +1312,58 @@ static void enh_conn_complete(struct bt_hci_evt_le_enh_conn_complete *evt)
...
 			atomic_clear_bit(bt_dev.flags, BT_DEV_ADVERTISING);
+
+			/*
+			 * There is no need to check ID address as only one
+			 * connection in slave role can be in pending state.
+			 */
+			conn = find_pending_connect(BT_HCI_ROLE_SLAVE, NULL);
+			if (!conn) {
+				BT_ERR("No pending slave connection");
+
```
Comment (jhedberg) [inline subsys/bluetooth/host/hci_core.c]: minor: remove this empty line

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/hci_core.c (hunk 3 above)
Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: Nevermind - I see that in the case conn is non-NULL the code jumps to the `done:` label where the unref happens

Comment (cvinayak) [reply subsys/bluetooth/host/hci_core.c]: `find_pending_connect()`  is now inside if-clause which have `goto done` that unref

**Review by jhedberg — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**